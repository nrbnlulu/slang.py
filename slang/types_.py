from __future__ import annotations
import abc
from dataclasses import dataclass
import dataclasses
import enum
from functools import cached_property
from typing import Callable, Iterable, TypeVar
from result import Result, Ok, Err


@dataclass(slots=True, kw_only=True)
class BaseField(abc.ABC):
    """Field in a record type."""

    name: str
    parent: BaseField | None = None

    @property
    def full_name(self) -> str:
        if self.parent is not None:
            return f"{self.parent.full_name}.{self.name}"
        return self.name

    @property
    def is_simple(self) -> bool:
        return False

    @property
    def is_namespace(self) -> bool:
        return False

    @property
    def is_complex(self) -> bool:
        return False

    @abc.abstractmethod
    def compatible(self, other: BaseField, current_locale: str) -> Result[None, str]:
        """Ensure that both of the fields are compatible."""
        pass


T_BaseField = TypeVar("T_BaseField", bound=BaseField)


def _find_field_error(
    field: T_BaseField,
    other: BaseField,
    locale_name: str,
    custom_error_mapper: Callable[[T_BaseField], str],
) -> str:
    if field.name != other.name:
        return f"{locale_name}: Field {field.full_name} is not compatible with {other.full_name}."
    if not isinstance(other, field.__class__):
        return f"{locale_name}: Field {field.full_name} is a {field.__class__.__name__}, but should be {other.__class__.__name__}."
    return custom_error_mapper(field)


@dataclass(slots=True, kw_only=True)
class NameSpaceField(BaseField):
    fields: list[BaseField] = dataclasses.field(default_factory=list)

    @property
    def is_namespace(self) -> bool:
        return True
    
    def get_child_namespace_fields(self) -> list[NameSpaceField]:
        ret: list[NameSpaceField] = []
        for field in self.fields:
            if isinstance(field, NameSpaceField):
                ret.extend(field.get_child_namespace_fields())
                ret.append(field)
        return ret

    @property
    def class_name(self) -> str:
        full_name = self.full_name
        return full_name.replace(".", "_").capitalize()

    @property
    def proto_name(self) -> str:
        return f"{self.class_name}Proto"

    def get_field(self, name: str) -> BaseField:
        for field in self.fields:
            if field.name == name:
                return field
        raise ValueError(f"Field {name} is not defined in {self.full_name}.")
    
    def compatible(self, other: BaseField, current_locale: str) -> Result[None, str]:
        """Ensure that both of the namespaces has the same fields."""
        if isinstance(other, NameSpaceField):
            for field in self.fields:
                if any(
                    field.compatible(other_field, current_locale)
                    for other_field in other.fields
                ):
                    return Ok(None)

        def pretty_error(other: NameSpaceField) -> str:
            error_str = f"{current_locale}: {self.full_name} is not compatible with {other.full_name}. \n"
            error_str += "The following fields are missing:"
            for field in self.fields:
                if not any(
                    field.compatible(other_field, current_locale)
                    for other_field in other.fields
                ):
                    error_str += f"\n{field.full_name}"
            return error_str

        return Err(_find_field_error(self, other, current_locale, pretty_error))


@dataclass(slots=True, kw_only=True)
class SimpleField(BaseField):
    value: str

    @property
    def is_simple(self) -> bool:
        return True

    def compatible(self, other: BaseField, current_locale: str) -> Result[None, str]:
        """Ensure that both of the fields are compatible."""
        if isinstance(other, SimpleField):
            if self.name == other.name:
                return Ok(None)
        return Err(
            _find_field_error(
                self,
                other,
                current_locale,
                lambda field: f"{current_locale}: Field {field.full_name} is not compatible with {other.name}.",
            )
        )


class ArgumentType(enum.Enum):
    INT = "int"
    STR = "str"
    FLOAT = "float"
    BOOL = "bool"

    @classmethod
    def from_string(cls, kind: str) -> ArgumentType:
        if kind == "int":
            return cls.INT
        if kind == "str":
            return cls.STR
        if kind == "float":
            return cls.FLOAT
        if kind == "bool":
            return cls.BOOL
        raise ValueError(f"Unknown argument kind: {kind}")

    @property
    def as_py(self) -> str:
        if self == ArgumentType.INT:
            return "int"
        if self == ArgumentType.STR:
            return "str"
        if self == ArgumentType.FLOAT:
            return "float"
        if self == ArgumentType.BOOL:
            return "bool"
        raise ValueError(f"Unknown argument kind: {self}")

@dataclass(slots=True, kw_only=True)
class ArgumentDefinition:
    name: str
    type: ArgumentType


@dataclass(slots=True, kw_only=True)
class ComplexField(BaseField):
    """Field which needs to render a string using arguments."""

    arguments: list[ArgumentDefinition]
    template: str

    @property
    def is_complex(self) -> bool:
        return True

    def get_argument(self, name: str) -> ArgumentDefinition:
        for arg in self.arguments:
            if arg.name == name:
                return arg
        raise ValueError(f"Argument {name} is not defined in {self.full_name}.")

    def compatible(self, other: BaseField, current_locale: str) -> Result[None, str]:
        if (
            isinstance(other, ComplexField)
            and self.name == other.name
            and self.arguments == other.arguments
        ):
            return Ok(None)

        def pretty_error(other: ComplexField) -> str:
            error_str = f"{current_locale}: {self.full_name} arguments is not compatible with {other.full_name}. \n"
            for arg in self.arguments:
                if arg not in other.arguments:
                    other_arg = other.get_argument(arg.name)
                    error_str += f"\n`{arg.name}: {arg.type}` != `{other_arg.name}: {other_arg.type}`"
            return error_str

        return Err(_find_field_error(self, other, current_locale, pretty_error))


@dataclass(slots=True, kw_only=True)
class LocaleImpl:
    """Context for a type checker."""

    name: str
    root: NameSpaceField | None = None

    def get_root(self) -> NameSpaceField:
        if self.root is None:
            raise ValueError("Root field is not defined.")
        return self.root

    def compatible(
        self, other: LocaleImpl, current_namespace: str
    ) -> Result[None, str]:
        """Ensure that both of the implementations has the same fields."""
        return self.get_root().compatible(other.get_root(), current_namespace)

    @property
    def all_namespace_fields(self) -> list[NameSpaceField]:
        ret: list[NameSpaceField] = []
        for field in self.get_root().fields:
            if isinstance(field, NameSpaceField):
                ret.extend(field.get_child_namespace_fields())
                ret.append(field)
        return ret
                
    

@dataclass(slots=True, kw_only=True)
class SlangCtx:
    ref_locale: LocaleImpl
    """Reference locale, all locales should match this one."""
    locales: list[LocaleImpl] = dataclasses.field(default_factory=list)
 