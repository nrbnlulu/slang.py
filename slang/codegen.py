from minijinja import Environment

from slang.types_ import SlangCtx, LocaleImpl

macros = \
"""
{% macro render_field_signature(field, return_stmt) -%}
{% if field.is_simple -%}
    @property
    def {{ field.name }}(self) -> str:
        {{ return_stmt }}
{% elif field.is_complex -%}
    def {{ field.name }}(self, 
    {%- for arg in field.arguments -%}
        {{ arg.name }}: {{ arg.type.as_py }}{% if not loop.last %}, {% endif %}
    {%- endfor -%}
    ) -> str:
        {{ return_stmt }}
{% elif field.is_namespace %}
    @property
    def {{ field.name }}(self) -> {{ field.proto_name }}:
        {{ return_stmt }}
{% endif -%}
{%- endmacro %}
"""

base_tr_proto = """
{%- from "macros" import render_field_signature %}
from __future__ import annotations
from typing import Protocol

{% for ns_field in locale.all_namespace_fields -%}
class {{ ns_field.proto_name }}(Protocol):

    {% for field in ns_field.fields -%}
    {{ render_field_signature(field, "raise NotImplementedError()") }}
    {% endfor %}
{% endfor -%}


class TranslationsProto(Protocol):
    {% for ns_field in locale.get_root().fields -%}
    {{ render_field_signature(ns_field, "raise NotImplementedError()") }}
    {% endfor %}

"""
def error_callback(x):
    raise ValueError(x)
tem_env = Environment(templates={"translations_proto": base_tr_proto, "macros": macros})


def render_translations_proto(ctx: SlangCtx) -> str:
    return tem_env.render_template("translations_proto", locale=ctx.ref_locale)