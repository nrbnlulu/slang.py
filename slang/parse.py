import re
from typing import Any
from slang.types_ import SlangCtx
from slang.types_ import BaseField, NameSpaceField, SimpleField, ArgumentDefinition, ComplexField


def parse(input: dict[str, Any]) -> SlangCtx:
    ctx = SlangCtx()
    ctx.root = parse_namespace(ctx, input, "root")
    return ctx

def parse_namespace(ctx: SlangCtx, input: dict[str, Any], name: str) -> NameSpaceField:
    ns: dict[str, BaseField] = {}
    for k, v in input.items():
        ns[k] = parse_entry(ctx, k, v)
    return NameSpaceField(name, ns)
complex_f_pattern = r'^(\w+)\((\w+: \w+)(, \w+: \w+)*\)$'

def parse_entry(ctx: SlangCtx, k: str, v: Any) -> BaseField:
    if match := re.match(complex_f_pattern, k):
        return parse_complex(ctx, match, v)
    if isinstance(v, dict):
        return parse_namespace(ctx, v, k)
    else:
        return SimpleField(k, v)


def parse_complex(ctx: SlangCtx, match: re.Match, v: str) -> ComplexField:
    name = match.group(1)
    args: list[ArgumentDefinition] = []
    for i in range(2, len(match.groups())):
        arg_name, arg_type = match.group(i).replace(" ", "").split(":")
        args.append(ArgumentDefinition(arg_name, arg_type))
    return ComplexField(name, args, template=v)
    
example = {
    'name(family: str)': 'John {family}',
    'age': 30,
    'address': {
        'street': '123 Main St',
        'city': 'Springfield',
        'state': 'IL',
        'zip': 62701
    },
    'phone': '555-1234',
    'email': "fdsaf@fadsf.com"
}
print(parse(example))