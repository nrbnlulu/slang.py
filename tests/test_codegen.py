from pathlib import Path
from slang.codegen import render_translations_proto
from slang.parse import parse
from slang.types_ import SimpleField, SlangCtx

def write_to_tmp(content) -> None:
    (Path(__file__).parent / "tmp.py").write_text(content)


def test_simple_field_proto(snapshot) -> None:
    example = {
        "simple": "Simple",
    }
    locale = parse(example, "heb").unwrap()
    assert snapshot == render_translations_proto(SlangCtx(ref_locale=locale))
    

def test_complex_field_proto(snapshot) -> None:
    example = {
        "complex(arg1: str, arg2: int, arg3: bool, arg4: float)": "Hello, {arg1}, {arg2}, {arg3}, {arg4}",
    }
    locale = parse(example, "heb").unwrap()
    assert snapshot == render_translations_proto(SlangCtx(ref_locale=locale))


def test_namespace_field_proto(snapshot) -> None:
    example = {
        "namespace": {
            "simple": "Simple",
            "complex(arg1: str)": "Hello, {arg1}",
        },
    }
    locale = parse(example, "heb").unwrap()
    write_to_tmp(render_translations_proto(SlangCtx(ref_locale=locale)))
    assert snapshot == render_translations_proto(SlangCtx(ref_locale=locale))