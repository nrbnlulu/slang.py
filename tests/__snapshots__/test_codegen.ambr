# serializer version: 1
# name: test_complex_field_proto
  '''
  
  from __future__ import annotations
  from typing import Protocol
  
  class TranslationsProto(Protocol):
      def complex(self,arg1: str, arg2: int, arg3: bool, arg4: float) -> str:
          raise NotImplementedError()
  
      
  
  '''
# ---
# name: test_namespace_field_proto
  '''
  
  from __future__ import annotations
  from typing import Protocol
  
  class Locale_heb_namespaceProto(Protocol):
  
      @property
      def simple(self) -> str:
          raise NotImplementedError()
  
      def complex(self,arg1: str) -> str:
          raise NotImplementedError()
  
      
  class TranslationsProto(Protocol):
      
      @property
      def namespace(self) -> Locale_heb_namespaceProto:
          raise NotImplementedError()
  
      
  
  '''
# ---
# name: test_simple_field_proto
  '''
  
  from __future__ import annotations
  from typing import Protocol
  
  class TranslationsProto(Protocol):
      @property
      def simple(self) -> str:
          raise NotImplementedError()
  
      
  
  '''
# ---
